### PR TITLE
feat: favorites view and API integration

### PR DESCRIPTION
**Pull request by SofiaCantero24**
## What does this do?
Favorites screen and API

https://github.com/user-attachments/assets/9ae068a2-817f-42d0-b023-38f5c92c9fb5

## Notion Ticket
* [Mark product as Favorite](https://www.notion.so/rootstrap/Mark-a-product-as-favorite-75b86970d59542d6979d1de1917c3693?pvs=4)

## Figma view
* [Favorites screen](https://www.figma.com/file/uUWXfDpuvDtnOHdSRjv0t1/Black-Market---iOS?type=design&node-id=4-714&mode=design&t=sIzqiHFLhyenWELW-4)

## How did you test this?

### CODE REVIEW

File: src/app/(app)/favorites.tsx [hunk 1]
```diff
@@ -1,21 +1,66 @@
...
+            renderItem={({ item, index }) => {
+              const isFirstItem = index === 0;
+              const isLastItem = index === productsToDisplay.length - 1;
+
+              return (
+                <View
+                  className={`mx-4 border ${
+                    isFirstItem ? 'rounded-t-lg' : ''
+                  } ${isLastItem ? 'rounded-b-lg' : ''}`}
+                >
```
Comment (Stukz) [inline src/app/(app)/favorites.tsx]: don't forget to use twMerge

File: src/app/(app)/product-list.tsx [hunk 2]
```diff
@@ -51,45 +52,49 @@ const SearchResult = ({
...
+          data={products}
+          renderItem={({ item, index }) => {
+            const isFirstItem = index === 0;
+            const isLastItem = index === products.length - 1;
+
+            return (
+              <View
+                className={`mx-4 border ${isFirstItem ? 'rounded-t-lg' : ''} ${
+                  isLastItem ? 'rounded-b-lg' : ''
+                }`}
```
Comment (Stukz) [inline src/app/(app)/product-list.tsx]: here also, use `twMerge`

File: src/components/favourites/favorite-card.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+const StateLabel = ({ state }: { state: string }) => {
+  return (
+    <View
+      className={`mb-2 mt-1 w-12 items-center space-x-2 rounded-md px-2 py-1 ${
+        state === 'used' ? 'bg-restored' : 'bg-new'
+      }`}
+    >
+      <Text className={`text-xs font-semibold text-white`}>
```
Comment (Stukz) [inline src/components/favourites/favorite-card.tsx]:
```suggestion
      <Text className="text-xs font-semibold text-white">
```

File: src/components/favourites/favorite-card.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  image_url: string;
+  refetch: () => void;
+}
+
+const StateLabel = ({ state }: { state: string }) => {
+  return (
+    <View
+      className={`mb-2 mt-1 w-12 items-center space-x-2 rounded-md px-2 py-1 ${
+        state === 'used' ? 'bg-restored' : 'bg-new'
+      }`}
```
Comment (Stukz) [inline src/components/favourites/favorite-card.tsx]: dont forget to use `twMerge`

File: src/components/products-list/product-card.tsx [hunk 5]
```diff
@@ -10,10 +19,85 @@ export interface ProductCardProps {
...
+    },
+  });
+  const { mutate: removeFavorite } = useRemoveFavorite({
+    onSuccess: () => {
+      refetch();
+    },
+    onError: () => {
+      showErrorMessage('Something went wrong');
+    },
+  });
```
Comment (guillermomachado) [inline src/components/products-list/product-card.tsx]: I feel this is too much responsability to the component to carry. Components should be mostly visual components and the logic should live in the parent

File: src/components/products-list/product-card.tsx [hunk 6]
```diff
@@ -10,10 +19,85 @@ export interface ProductCardProps {
...
+
+const TitleAndFavorite = ({
+  id,
+  name,
+  isFavorite,
+  refetch,
+}: TitleAndFavoriteProps) => {
+  const { mutate: addFavorite } = useAddFavorite({
+    onSuccess: () => {
+      refetch();
```
Comment (guillermomachado) [inline src/components/products-list/product-card.tsx]: I recommend you to read about queries invalidations: https://tanstack.com/query/v3/docs/framework/react/guides/invalidations-from-mutations so you don't need to care about manually refetching.

File: src/components/favourites/favorite-card.tsx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            className="flex h-28 w-24"
+            contentFit="contain"
+          />
+        </Link>
+        <View className="w-full flex-1 px-4 py-3">
+          <View className="flex-row items-center justify-between">
+            <Link
+              href={{
+                pathname: detailsRoute,
+                params: { id: id },
```
Comment (Stukz) [inline src/components/favourites/favorite-card.tsx]:
```suggestion
                params: { id },
```

File: src/components/favourites/favorite-card.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    refetch,
+  }: ShoppingProductCardProps) => {
+    const { mutate: removeFavorite } = useRemoveFavorite({});
+    const detailsRoute = '/details/[id]';
+    return (
+      <View className="w-full flex-row items-center justify-between space-x-4 rounded-lg bg-white p-3">
+        <Link
+          href={{
+            pathname: detailsRoute,
+            params: { id: id },
```
Comment (Stukz) [inline src/components/favourites/favorite-card.tsx]:
```suggestion
            params: { id },
```

File: src/api/favorites/add-favorite.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+export const useAddFavorite = (
+  props: UseMutationOptions<FavoriteProduct, string, number, any>
+) => {
+  const queryClient = useQueryClient();
+
+  return useBaseMutation<number, FavoriteProduct>({
+    mutationKey: ['addFavorite'],
+    mutationFn: addFavorite,
+    onSuccess: (data, variables, context) => {
+      queryClient.invalidateQueries({ queryKey: ['products'] });
```
Comment (guillermomachado) [inline src/api/favorites/add-favorite.ts:26]: Cool!

**Review by guillermomachado — APPROVED** [review]

**Review by Stukz — APPROVED** [review]

**Pull request merged**